File: beginner-dsa/solutions/python/Q_2_min_stack.py
class MinStack:
    def __init__(self):
        self.stack = []      # Main stack to store values
        self.min_stack = [float('inf')]  # Min stack initialized with infinity as sentinel

    def push(self, val: int) -> None:
        self.stack.append(val)
        # Push current minimum between val and last min onto min_stack
        current_min = min(val, self.min_stack[-1])
        self.min_stack.append(current_min)

    def pop(self) -> None:
        if self.stack:
            self.stack.pop()
            self.min_stack.pop()

    def top(self) -> int:
        if self.stack:
            return self.stack[-1]
        return None

    def getMin(self) -> int:
        # The last element in min_stack is current minimum
        if self.min_stack:
            return self.min_stack[-1]
        return None

# Example usage:
min_stack = MinStack()
min_stack.push(-2)
min_stack.push(0)
min_stack.push(-3)

print(min_stack.getMin())  # Returns -3
min_stack.pop()
print(min_stack.top())     # Returns 0
print(min_stack.getMin())  # Returns -2

File: beginner-dsa/solutions/cpp/Q2_min_stack.cpp
#include <iostream>
#include <stack>
using namespace std;

class MinStack {
private:
    stack<int> s;        // main stack
    stack<int> minStack; // stack to keep track of minimum elements

public:
    // Initialize the stack object
    MinStack() {}

    // Push element onto stack
    void push(int val) {
        s.push(val);
        // If minStack is empty or val <= current min, push onto minStack
        if (minStack.empty() || val <= minStack.top()) {
            minStack.push(val);
        }
    }

    // Remove the element on top of stack
    void pop() {
        if (s.empty()) return;
        int topVal = s.top();
        s.pop();
        // If the popped value is the current minimum, pop from minStack too
        if (!minStack.empty() && topVal == minStack.top()) {
            minStack.pop();
        }
    }

    // Get the top element
    int top() {
        if (s.empty()) {
            throw runtime_error("Stack is empty");
        }
        return s.top();
    }

    // Retrieve the minimum element in the stack
    int getMin() {
        if (minStack.empty()) {
            throw runtime_error("Stack is empty");
        }
        return minStack.top();
    }
};

// Driver code
int main() {
    MinStack st;
    st.push(-2);
    st.push(0);
    st.push(-3);
    cout << "Current minimum: " << st.getMin() << endl; // Output: -3
    st.pop();
    cout << "Top element: " << st.top() << endl;        // Output: 0
    cout << "Current minimum: " << st.getMin() << endl; // Output: -2
    return 0;
}
